internal/dataplane/kong_client.go

```diff
@@ -489,14 +492,21 @@ func (c *KongClient) Update(ctx context.Context) error {
 
 	// In case of a failure in syncing configuration with Gateways, propagate the error.
 	if gatewaysSyncErr != nil {
-		if recoveringErr := c.tryRecoveringFromGatewaysSyncError(ctx, cacheSnapshot, gatewaysSyncErr); recoveringErr != nil {
+		if recoveringErr := c.tryRecoveringFromGatewaysSyncError(
+			ctx,
+			cacheSnapshot,
+			gatewaysSyncErr,
+		); recoveringErr != nil {
 			return fmt.Errorf("failed to recover from gateways sync error: %w", recoveringErr)
 		}
 		// Update result is positive only if gateways were successfully synced with the current config, so we still
 		// need to return the error here even if we succeeded recovering.
 		return gatewaysSyncErr
 	}
 
+	// if we've gotten here we've pushed a successful config, so purge any existing fallback diagnostic info
+	c.diagnostic.Fallbacks <- diagnostics.FallbackDiagnosticCollection{Objects: []diagnostics.FallbackDiagnostic{}}
```

Comment:
1. When diagnostics server [is disabled](https://github.com/Kong/kubernetes-ingress-controller/blob/cd9b41c2d53927dff3a0754746f670fa7ad0e966/internal/cmd/rootcmd/servers.go#L19-L22) this will block, right?
2. To make context cancellation work, this should use a `select` on `ctx.Done()` and this channel:

```suggestion
	select {
	case <-ctx.Done():
		return ctx.Err():
	case c.diagnostic.Fallbacks <- diagnostics.FallbackDiagnosticCollection{Objects: []diagnostics.FallbackDiagnostic{}}
	}
```

Reply:
a329ccb